### ISSUE TITLE
[Flang] Incorrect diagnose on SELECT TYPE type guard statement when the selector is of a PDT type.

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#83612]
Consider the following code:
```
  MODULE M

    TYPE :: Base(K1,N1)    ! (4,20)
      INTEGER, KIND :: K1
      INTEGER, LEN  :: N1
    END TYPE

    TYPE, EXTENDS(Base)  :: DT    ! (4,20)
      PROCEDURE(ModFun2), PASS, POINTER :: ProcPtr2
    END TYPE

    CONTAINS

    FUNCTION ModFun2(Arg)
    CLASS(DT(4,*)) :: Arg
    CLASS(DT(4,:)), ALLOCATABLE :: ModFun2
    END FUNCTION

  END MODULE


  PROGRAM Final2
  USE M
  TYPE(DT(4,:)), ALLOCATABLE  :: V

  ALLOCATE(v, SOURCE=DT(4,20)(ModFun2))

  SELECT TYPE ( As =>  v%ProcPtr2() )
  type IS (DT(4,*))
  END SELECT

  END
```

Flang currently issues an error:
```
:29:12: error: Type specification 'dt(k1=4_4,n1=*)' must be an extension of TYPE 'dt(k1=4_4,n1=:)'
    type IS (DT(4,*))
             ^^^^
```

The code seems conforming to me.

**Comment by klausler** [comment]
https://github.com/llvm/llvm-project/pull/83888

**Comment by DanielCChen** [comment]
I have verified the reproducer and the original test cases are fixed by PR #83888.
